### ISSUE TITLE
Preserve line/paragraph breaks in description from Template Generator

### ISSUE THREAD

**Issue by twinkietoes-on** [LibriVox/librivox-catalog#188]
A reader fills in the Template Generator, adding paragraphs to the description. These show as paragraph breaks in the database entry, but they're lost at cataloging, both on the catalog page and on the data sent to Archive.org.
![Capture1](https://github.com/LibriVox/librivox-catalog/assets/70523393/c9fdf965-babe-46f2-abef-c90bc959cea6)
![Capture2](https://github.com/LibriVox/librivox-catalog/assets/70523393/de72180a-2328-46d1-a91c-bf1879ff2095)

Current fix is to manually enter `<br />` html to the database field. (Paragraph tags cause other problems in the display.)

**Comment by twinkietoes-on** [comment]
If this involves editing the code that goes to Archive, I'd like to see a couple additional line breaks in there, between the different boilerplate components (top and bottom) and the description, to separate them into paragraphs as well. That's not as critical, but a "might as well, while you're in there" item.

**Comment by garethsime** [comment]
Getting all the newlines to show in the catalog view is a one-line change, but it presents an interesting question around what to do with descriptions that are already using `<br />` tags. For example, there might be entries like this in the management view:

> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec tincidunt auctor ligula. Donec eu lacinia metus. Vestibulum vel felis vel sapien condimentum ullamcorper eget eget massa. Vestibulum pharetra purus eu erat imperdiet, vitae dignissim massa malesuada. Curabitur sit amet magna quis elit tristique fringilla. &lt;br /&gt;&lt;br /&gt;
>
> Quisque id lorem leo. Sed iaculis nisl lectus, quis congue leo blandit in. Vivamus facilisis velit ut metus fermentum, ut tristique ex sodales. Proin gravida viverra velit, sed finibus dui mollis a. 

Which would then end up being rendered with extra breaks like this:

> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec tincidunt auctor ligula. Donec eu lacinia metus. Vestibulum vel felis vel sapien condimentum ullamcorper eget eget massa. Vestibulum pharetra purus eu erat imperdiet, vitae dignissim massa malesuada. Curabitur sit amet magna quis elit tristique fringilla.
> &nbsp;
> &nbsp;
>
> Quisque id lorem leo. Sed iaculis nisl lectus, quis congue leo blandit in. Vivamus facilisis velit ut metus fermentum, ut tristique ex sodales. Proin gravida viverra velit, sed finibus dui mollis a.

My copy of the scrubbed database is probably a year old by now, but it's got several thousand projects that have `br` tags (both kinds) followed by newline/return characters. 

Perhaps we could do something clever where we only do the replacement if there are no `br` tags in the description, or we could try to collapse together `br` tags with adjacent whitespace, but that seems like it could lead to confusion about when the `br` tags are needed and when they aren't.

Thoughts?

**Comment by garethsime** [comment]
Here's a cool idea -- https://www.darklaunch.com/php-normalize-newlines-line-endings-crlf-cr-lf-unix-windows-mac.html

They normalise all of the line endings and then check that you don't get more than two in a row. It might not be 100% what people intended, but I think it would be pretty close, and it would prevent any of the descriptions from having massive spaces like I was suggesting in the last comment.

**Comment by twinkietoes-on** [comment]
That sounds like a good solution. Could it go on the test server, to be sure it's doing what we want it to do?

**Comment by redrun45** [comment]
Looks like this went straight to live. Seems to work as advertised, though. 
I can update the [Cast Coder](https://ia800702.us.archive.org/31/items/LibrivoxM4bCollectionAudiobooks/CastCoder.html) to leave out the `<br>` tags after each role now.  I'll shoot for this weekend on that.

@twinkietoes-on, any other special cases come to mind?
This change doesn't affect what's sent to Archive, only the display on our catalog pages.  Would you like as a separate issue?

**Comment by notartom** [comment]
> Looks like this went straight to live.

That's on me - figured it's better to ask for forgiveness than permission, in this case.

**Comment by redrun45** [comment]
Cast coder has been updated: it separates entries with `\n` newlines rather than with `<br>` tags.  The old version has just been renamed, since that little project doesn't have git to keep its history.

Guess we may as well keep this issue open for the Archive text.  50% down, 50% to go!  :grinning:

**Comment by redrun45** [comment]
Thank you @garethsime, that was extremely helpful!  I've switched the controller to calling your helper instead of `trim`, and I think that will do it.  :grin: 

@twinkietoes-on, I've also found the template that includes the other spacing elements.  I'm adding another break just before and just after the description.  :+1: 

I'll see if I can put together a unit test to cover this - I've never done one of those before!  And then, I'll see if I can figure out credentials and catalog a few dummy projects to be sure.  On a dummy _account_, if I can.  :laughing:

**Comment by twinkietoes-on** [comment]
Sounds great!

    On Sunday, March 3, 2024 at 11:46:26 a.m. EST, redrun45 ***@***.***> wrote:  
 
 


Thank you @garethsime, that was extremely helpful! I've switched the controller to calling your helper instead of trim, and I think that will do it. 😁

@twinkietoes-on, I've also found the template that includes the other spacing elements. I'm adding another break just before and just after the description. 👍

I'll see if I can put together a unit test to cover this - I've never done one of those before! And then, I'll see if I can figure out credentials and catalog a few dummy projects to be sure. On a dummy account, if I can. 😆
